src/net/server/message.rs

```diff
@@ -0,0 +1,402 @@
+    }
+
+    /// From which IP address and port number was this message received?
+    pub fn client_addr(&self) -> std::net::SocketAddr {
+        self.client_addr
+    }
+
+    /// Read access to the inner message
+    pub fn message(&self) -> &T {
+        &self.message
+    }
+}
+
+//--- Clone
+
+impl<T> Clone for Request<T> {
+    fn clone(&self) -> Self {
+        Self {
+            client_addr: self.client_addr,
+            received_at: self.received_at,
+            message: Arc::clone(&self.message),
+            transport_specific: self.transport_specific,
+        }
+    }
+}
+
+//----------- MessageProcessor -----------------------------------------------
+
+pub struct MessageDetails<Buf>
+where
+    Buf: BufSource,
+{
+    buf: Buf::Output,
+    received_at: Instant,
+    addr: SocketAddr,
+}
+
+impl<Buf> MessageDetails<Buf>
+where
+    Buf: BufSource,
+{
+    pub fn new(
+        buf: Buf::Output,
+        received_at: Instant,
+        addr: SocketAddr,
+    ) -> Self {
+        Self {
+            buf,
+            received_at,
+            addr,
+        }
+    }
+}
+
+/// Perform processing common to all messages being handled by a DNS server.
+///
+/// All messages received by a DNS server need to pass through the following
+/// processing stages:
+///
+///   - Pre-processing.
+///   - Service processing.
+///   - Post-processing.
+///
+/// The strategy is common but some server specific aspects are delegated to
+/// the server that implements this trait:
+///
+///   - Adding context to a request.
+///   - Finalizing the handling of a response.
+///
+/// Servers implement this trait to benefit from the common processing
+/// required while still handling aspects specific to the server themselves.
+///
+/// Processing starts at [`process_request()`].
+///
+/// [`process_request()`]: Self::process_request()
+pub trait MessageProcessor<Buf, Svc>
+where
+    Buf: BufSource + Send + Sync + 'static,
+    Buf::Output: Octets + Send + Sync + 'static,
+    Svc: Service<Buf::Output> + Send + Sync + 'static,
+{
+    type State: Clone + Send + Sync + 'static;
+
+    /// Process a DNS request message.
+    ///
+    /// This function consumes the given message buffer and processes the
+    /// contained message, if any, to completion, possibly resulting in a
+    /// response being passed to [`Self::process_call_result()`].
+    ///
+    /// The request message is a given as a seqeuence of bytes in `buf`
+    /// originating from client address `addr`.
+    ///
+    /// The [`MiddlewareChain`] and [`Service`] to be used to process the
+    /// message are supplied in the `middleware_chain` and `svc` arguments
+    /// respectively.
+    ///
+    /// Any server specific state to be used and/or updated as part of the
+    /// processing should be supplied via the `state` argument whose type is
+    /// defined by the implementing type.
+    ///
+    /// On error the result will be a [`ServiceError`].
+    fn process_request(
```

Comment:
`MessageProcessor::process_request()` is not supposed to ever be overridden by a custom implementation. In fact `MessageProcessor` shouldn't even be a trait. I discussed this with @partim a few weeks ago and we tried then to replace it without using a trait but ran into difficulties.

Essentially this trait is being abused as an easy way to share common logic between the TCP and UDP server implementations, and is an internal detail of the library. It may be used for another future server impl, provided by us or a library user, but should be used as-is, its default function impls are not intended to be replaced.

So, yes, I think this is wrong. And yes we want to change it. But no we haven't yet worked out how to do so.

I wouldn't go as far as calling it risky however.